torch/_functorch/_aot_autograd/runtime_wrappers.py

```diff
@@ -1331,3 +1347,608 @@ def _same_dtype_views(view1, view2):
+    Got grad_output types: {str(grad_output_types)}"""
+
+                # TODO: figure out how to refactor the backward properly
+                # so I can use aot_dispatch_subclass_wrapper() here.
+                if CompiledFunction.maybe_subclass_metadata is not None:
+                    # Get the number of tangents after unwrapping
+                    len_tangents = len(
+                        unwrap_tensor_subclasses(
+                            all_args[tangents_start_idx:tangents_end_idx],
+                            is_joint_structure=False,
+                        )
+                    )
+                    all_args = [
+                        AOTDispatchAutograd.coerce_runtime_tangent(
+                            t,
+                            CompiledFunction.metadata.traced_tangents[
+                                i - tangents_start_idx
+                            ],
+                        )
+                        if tangents_start_idx <= i < tangents_end_idx
+                        else t
+                        for i, t in enumerate(all_args)
+                    ]
+                    all_args = unwrap_tensor_subclasses(
+                        all_args, is_joint_structure=False
+                    )
+                    tangents_start_idx = len(all_args) - len_tangents - len(rng_args)
+                    tangents_end_idx = tangents_start_idx + len_tangents
+
+                # Make the tangents contiguous. Note that we must do this after subclass desugaring
+                # because inputs to inductor have to be contiguous
+                all_args = [
+                    AOTDispatchAutograd._force_contiguous(t)
+                    if (tangents_start_idx <= i < tangents_end_idx)
+                    else t
+                    for i, t in enumerate(all_args)
+                ]
+
+                def call_compiled_backward():
+                    if ctx._is_compiled_autograd_tracing():
+                        assert lazy_backward_info is not None
+                        bw_module = lazy_backward_info.bw_module
+                        # For compiled autograd, run raw FX graph so that it can be inlined into the larger graph
+                        symints = ctx._get_compiled_autograd_symints()
+                        assert len(symints) == len(ctx.symints)
+                        all_args[: len(symints)] = symints
+                        if backward_state_indices:
+                            assert (
+                                ctx._compiled_autograd_backward_state.proxy is not None
+                            )
+                            all_args.append(ctx._compiled_autograd_backward_state)
+                        context = (
+                            torch._C._DisableAutocast if disable_amp else nullcontext
+                        )
+                        with context():
+                            out = normalize_as_list(bw_module(*all_args))
```

Comment:
V0 of my implementation assumes no compiled autograd. But yes, I think if we want to support it we would need to serialize the bw_module. 

Note that serializing a graph module today changes the node's names and does a bunch of other minor tweaks. I feel like serializing it for just this use case is fine if we can make some invariants about exactly what compiled autograd needs to trace through and make sure those are always preserved. 

Specifically, is it true that we will know whether compiled autograd is turned on at forward compile time? That is, you shouldn't have a situation where compiled autograd is off when compiling the forward and on when compiling the backward, right? If so, we could just save the bw_module to the cache only when compiled_autograd is turned on, and then use that here as needed. 

Let's discuss this more IRL, I think it's pretty confusing to write over text, but it sounds to me like these are just more guards you'd need to implement, because they're cache keys you can't compute at compile time.

Reply:
So first off, we absolutely should not let compiled autograd get in the way of a useful cache for non-compiled autograd cases.

That being said, the compiled autograd situation is very interesting.

Let's considered a standard use case for compiled autograd, which is a Python backward hook that needs to be Dynamo traced through at backwards time. With no caches, what we do is we generate an unserializable backwards graph that can contain arbitrary function calls to arbitrary Python code. On its face, this is unserializable, because we can't serialize arbitrary Python functions.

However, at the point in time when we do Dynamo tracing in compiled autograd, there is a sense in which where we *can* serialize arbitrary Python functions, in the same sense that we could implement a Dynamo-level cache: if we ensure all the code objects are the same, and we ensure all the guards match, we know exactly what path we took through the serialized Python function, and we don't actually need to serialize the Python function (because we're going to go to a compiled product.)

However, this is a bit unsatisfactory. In particular, what if the Python function has a graph break, or it ends up needing to be recompiled on cache hit? Then you really do need the original Python function and you're back to square one.

So maybe you do need some strategy for serializing arbitrary Python functions to get this to work. Very very delicate...